### ISSUE TITLE
Where is the good performance from version 0.76 ?

### ISSUE THREAD

**Issue by RaymanXY** [dosbox-staging/dosbox-staging#1482]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 10

### Whats your question and how can we help?

Since version 0.76 I'm trying to upgrade to the newer version 0.77.1 and 0.78. Since I'm still having the same issue with 0.78, I decided now to post this issue here.

**What's wrong?**
- 0.76 (32bit): Super performance, high frame rates, no stuttering. Everything works like a charm.
- 0.77.1 / 0.78 (tried out 32bit and 64bit): Worse performance, longer loading times, lower frame rates, stuttering

**Where is this happening?**
Only when running certain games. Most of the games are running well even with the latest version. Affected games are:

- **Flight Unlimited**: Almost unplayable with the new version in 32bit. It's better with 64bit, but still stuttering. With 0.76 32bit I'm having no issues at all.

- **Manic Karts**: Needs much more time to load a race and some noticable frame rate issues in-game. Here I've captured a video from both versions. You'll see the much longer loading times:
[0.76.zip](https://github.com/dosbox-staging/dosbox-staging/files/7794738/0.76.zip)
[0.78.zip](https://github.com/dosbox-staging/dosbox-staging/files/7794739/0.78.zip)

I'm using the same config file. So I just did an upgrade to the newer version and typed in `config -wc`.
My settings are attached to this ticket, see [dosbox-staging 0.76.zip](https://github.com/dosbox-staging/dosbox-staging/files/7794678/dosbox-staging.zip) and [dosbox-staging 0.78.zip](https://github.com/dosbox-staging/dosbox-staging/files/7794828/dosbox-staging.0.78.zip)

**What have you tried out?**
I tried out different settings for `core` and `output`, nothing helped.

**Computer Hardware**
- Lenovo LEGION Y540 Notebook
- Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
- 16 GB RAM
- NVIDIA GeForce GTX 1660 Ti

**Questions**
1. Is that issue related to my computer? Then I will accept it and stay on 0.76.
2. Does anyone have the same experience?
3. What could cause the poorer performance, so what happend between 0.76 and 0.77.1 and still not fixed in 0.78?
4. If you would like to investigate this issue, can I assist with logfiles?

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by kcgen** [comment]
Thanks for the report @RaymanXY; one notable difference in these newer versions is vsync (which should default to 'off').  
There's also a frame timeout warning log; do you notice orange/yellow frame timeout warnings filling the console output when you're inside a game?

**Comment by RaymanXY** [comment]
That's strange, I have only managed it once when I was not in fullscreen, see [console 0.78 windowed.txt](https://github.com/dosbox-staging/dosbox-staging/files/7794832/console.0.78.windowed.txt)

But I can not reproduce it anymore. Now it looks like [console 0.78.txt](https://github.com/dosbox-staging/dosbox-staging/files/7794835/console.0.78.txt). But even when the console remains empty without these warnings, the issue is still there...

**Comment by kcgen** [comment]
@RaymanXY , when you have a chance to try, here's builds with a potential fix: https://github.com/dosbox-staging/dosbox-staging/actions/runs/1639458078

**Comment by RaymanXY** [comment]
OK, that's an improvement! I've tested both, 32bit and 64bit.

**0.79 Alpha 32bit**
- Manic Karts: Problem solved
- Flight Unlimited: Still worse performance and low frame rates

**0.79 Alpha 64bit**
- Manic Karts: Problem solved
- Flight Unlimited: Better performance now, it's playable, but still not as good as 0.76 (32bit). It feels like in 0.76 there are still more fps and it's running smoother. The difference is not that big, but still noticable.

There is no difference between using my external 144hz Monitor or using my internal notebook screen, which is running at 144hz too. In the NVIDIA Control Panel I configured a maximum of 100 FPS. Don't know if these information are relevant.

I think Flight Unlimited needs more CPU power (?) and the improvement is not enough to reach the good 0.76 performance on my computer. If there is anything I can do or test, please let me know. It would make me very happy, if we reach the same performance as in 0.76.

**Comment by kcgen** [comment]
> * still not as good as 0.76 (32bit). It feels like in 0.76 there are still more fps

Do the MSVC builds make any difference?

https://github.com/dosbox-staging/dosbox-staging/actions/runs/1639477442

**Comment by Kappa971** [comment]
A few months ago, when the `vesa_refresh_hz` function was introduced, I tested it with Quake (I don't usually use Quake in DOSBox). I read on the internet that Quake runs at 72hz. Having a Freesync monitor I was able to do some tests. By setting `vesa_refresh_hz = 70` (default), I had some stuttering in the game... the game was producing 70fps but it's like it's still set at 72hz. In fact I solved the problem by setting `vesa_refresh_hz = 72`. Now I'm not an expert, I don't know what all this means, but I seem to have read here that VESA games should adapt to the monitor refresh rate, but it seems that with DOSBox the opposite happens.

@RaymanXY do you have freesync monitors? I have read many opinions on the internet about the proper functioning of these monitors. I believe the most truthful and logical (and that's what I use), is to remove any fps limit set and enable vsync from the NVIDIA control panel for everything (and disable it in games), which will act as an fps limiter in case the game exceeds 144 fps. In support of this thesis I noticed that, in fact, in some games there was tearing in the lower part of the screen if the vsync was not active.

**Comment by Grounded0** [comment]
Do we have LTO enabled in MSVC builds? That would explain a 10% loss of performance.

**Comment by RaymanXY** [comment]
> Do the MSVC builds make any difference?
> 
> https://github.com/dosbox-staging/dosbox-staging/actions/runs/1639477442

No, I've tested both, 32bit and 64bit and I've noticed no difference.


> @RaymanXY do you have freesync monitors?

Yes, ViewSonic XG2402. I tried removing the fps limit, enable / disable G-SYNC (all in NVIDIA Control Panel), but I've noticed no difference at all.

Can I mesure the fps in-game to share this with you guys?

**Comment by Grounded0** [comment]
[PERFDOS.zip](https://github.com/dosbox-staging/dosbox-staging/files/7796151/PERFDOS.zip)

Many of us use this to check for performance differences between builds here.

**Comment by kcgen** [comment]
> Do we have LTO enabled in MSYS2 builds? That would explain around 10% loss of performance.

Not for 0.78.x's current MSYS2 GCC-based release builds, but @shermp had it queued up for for merge in the 0.79.x series.

@Kappa971 - this is an interesting suggestion! Thanks for sharing it here. I'm looking at the VESA rates here: http://tinyvga.com/vga-timing , I'm curious if you can try some of the other specific VESA resolution & refresh rates combos with Quake.
If it can tolerate them (ie: no strange flickering), then we could add:
```ini
[VESA]
640x480_refresh = ...
800x600_refresh = ...
1024x768_refresh = ...
```
With specific choices for each based on the VESA tables.

**Comment by Kappa971** [comment]
I tried with DOSBox 0.78.0... with Quake I think it behaves in the correct way as 72fps are produced but there is a lot of stuttering, it was not present in the build tested months ago where you could manually set the `vesa_refresh_hz` option. With this https://github.com/dosbox-staging/dosbox-staging/actions/runs/1639477442 instead we have returned to the previous behavior where all the games run at 70 fps, the stuttering is less but perhaps this is not the behavior correct.

**Comment by RaymanXY** [comment]
> [PERFDOS.zip](https://github.com/dosbox-staging/dosbox-staging/files/7796151/PERFDOS.zip)
> 
> Many of us use this to check for performance differences between builds here. I have checked against this since at least 0.76.0 and it has been around 430 for my system on macOS since then.

I've tested this with 0.76 32bit (running smooth) and 0.78.1 32bit (stuttering in-game). BUT, this test gaves me the same result for both, around 400 for my system.

**Comment by kcgen** [comment]
Thanks for all the testing!

 @RaymanXY,
Can you run flight unlimited in Window mode, get to he parts where it stutters/or not as smooth as before. Then save your entire console log?

**Comment by Kappa971** [comment]
@kcgen, what is the first build where vesa_refresh_hz was introduced (in the DOSBox config file)? From what I remember, manually setting vesa_refresh_hz = 72, Quake was smooth (if it maintained 72 fps) but I wanted to be sure.

**Comment by RaymanXY** [comment]
> @RaymanXY, Can you configure DOSBox with `[mixer] nosound=true`, run flight unlimited in Window mode, get to the part where it's not as smooth as 0.76, and then post your console log?

Sure! Here are the console logs (nosound for both). I've taken these two versions because then we compare 32bit with 32bit. Let me know if I should compare other versions.
- 0.78.1 32bit (heavy stuttering) => [console 0.78.1 32bit nosound.txt](https://github.com/dosbox-staging/dosbox-staging/files/7796558/console.0.78.1.32bit.nosound.txt)
- 0.76 32bit (very smooth, although there are errors in the log) =>[console 0.76 32bit nosound.txt](https://github.com/dosbox-staging/dosbox-staging/files/7796559/console.0.76.32bit.nosound.txt)

Regarding reproduce and settings: I'll edit this post and add this information a bit later.

**Comment by Kappa971** [comment]
> @Kappa971 - this is an interesting suggestion! Thanks for sharing it here. I'm looking at the VESA rates here: http://tinyvga.com/vga-timing , I'm curious if you can try some of the other specific VESA resolution & refresh rates combos with Quake. If it can tolerate them (ie: no strange flickering), then we could add:
> 
> ```ini
> [VESA]
> 640x480_refresh = ...
> 800x600_refresh = ...
> 1024x768_refresh = ...
> ```
> 
> With specific choices for each based on the VESA tables (with defaults set to conservative known-good values).

I only tried 640x480, with higher resolutions I can't reach 72 fps, you should fix the stuttering issues first. What DOSBox 0.78.0 does I think is useful because it sets vesa_refresh_hz automatically. The problem is that with this build it's not smooth anyway.

**Comment by kcgen** [comment]
### Staging 0.76.0
 - In Microsoft's checksum-based allowlist
 - No realtime instructions scanning performed

![Capture-flight-760](https://user-images.githubusercontent.com/1557255/147855337-3c272449-d6d7-44f7-8f4c-723956875288.PNG)

### Staging 0.78.1 (realtime-scanning enabled)
 - binary and dll's not in Microsoft' checksum-based allowlist (yet)
 - All dynamic-core instructions are processed by the realtime scanner
 - All host CPU cores maxed out

![Capture-flight-781-msvc_realtime_virus_scan](https://user-images.githubusercontent.com/1557255/147855335-1f8fdc70-dcf9-4cb7-a1a3-2edf37719e2e.PNG)

### Staging 0.78.1 (realtime-scanning disabled)
 - All dynamic-core instructions run without burden

![Capture-flight-781](https://user-images.githubusercontent.com/1557255/147855333-d9f67173-9b2d-4f68-93e6-30a8f4552a0a.PNG)

For now, toggling RT scanning off on Windows is a (terrible) workaround.

Does anyone know how to submit a package to Microsoft and get in their checksum allow-list?  This has always been a problem for "new" binaries (and with time they are moved to the allow list).  But it would be nice to submit new packages on day one.

**Comment by RaymanXY** [comment]
> For now, toggling RT scanning off on Windows is a (terrible) workaround.

I was wondering if I'm able to apply this workaround on my system, just to check if the issue has been solved on my system. But I'm running Norton Security and Windows Defender is switched off.

Comparing the two versions regarding CPU load, I can not see any difference on my system:

**0.76 32bit (running smooth)**

![0 76](https://user-images.githubusercontent.com/63668523/147858064-510b103c-94e2-4302-baa3-fc1f603f3915.png)

**0.78.1 32bit (heavy stuttering)**

![0 78 1](https://user-images.githubusercontent.com/63668523/147858067-aa185092-acaf-40aa-bfb4-b813fa5e2e4c.png)

**Comment by kcgen** [comment]
Interesting. 

In your `C:\Users\Daniel\AppData\Local\DOSBox\dosbox-staging-FLIGHT.conf`, if you set a fixed number of cycles:

```ini
[cpu]
cycles = 180000
```

Curious how those compare in-game and CPU-usage wise.

**Comment by RaymanXY** [comment]
OK, I set the following in `dosbox-staging-FLIGHT.conf`:

`[cpu]`
`cycles = 180000`

`[autoexec]`
`MOUNT C ..\GAMES\FLIGHT`
`IMGMOUNT D C:\ISO\FL141DS.ISO -t cdrom`
`C:`
`FLIGHT.BAT`

**0.76 32bit (running smooth)**

![0 76 180000 cycles](https://user-images.githubusercontent.com/63668523/147858726-5a282c9c-9870-489d-b6f1-eda435541b83.png)


**0.78.1 32bit (ATTENTION: stuttering is a bit less now, but still not smooth)**

![0 78 1 180000 cycles](https://user-images.githubusercontent.com/63668523/147858732-bd2aee63-ca31-435d-98e7-f002ad2e5856.png)

**Comment by kcgen** [comment]
.. there's still something very different going on, and it's not jumping out at me yet.

We're both running Windows 10. We're both running the same conf files in our A-vs-B comparisons (so cycles, mixer blocksize, prebuffer, etc are all the same between side-by-side runs).

On your system, at 180k cycles:
 - 0.76 consumes ~60% of one core@2.13 Ghz: GPU usage is 8%
 - 0.78.x consumes 100%+ of one core@3.68 Ghz: GPU usage is only 6% (expected, fewer FPS)

In my case (below), Intel NUC:
 - 0.76 consumes ~100% of one core@2.41 Ghz: GPU usage is only 17%
 - 0.78.x consumes ~53%+ (bouncing around)@2.5 Ghz; GPU usage is 19%

The big difference is that your 0.78 saturates one core at 3.68 Ghz (and still wants more to satisfy 180k cycles), where as my (lower performance CPU and embedded intel GPU?) seems to be able to satisfy 180k cycles at about 50% total CPU usage at 2.5 Ghz.  I would also say my framerates are very smooth.. much more so than both of the video captured on your system. 

(For comparison on Linux and macOS, performance is very smooth without audio dropouts, all the way around cycles=300k -and I can't see any difference versus 0.76 -  CPU usages are framerates are the same) 

#0.76.0

![Capture-flight-760](https://user-images.githubusercontent.com/1557255/147855337-3c272449-d6d7-44f7-8f4c-723956875288.PNG)

### 0.78.1 (Window "defender" RT scanning toggled off)

![Capture-flight-781](https://user-images.githubusercontent.com/1557255/147855333-d9f67173-9b2d-4f68-93e6-30a8f4552a0a.PNG)

**Comment by RaymanXY** [comment]
> On your system, at 180k cycles:
>
>     * 0.76 consumes ~60% of one core@2.13 Ghz: GPU usage is 8%
> 
>     * 0.78.x consumes 100%+ of one core at a whopping 3.68 Ghz but still produces fewer FPS (6% GPU usage)

I just re-tested 0.78.1 again to verify this whopping. I think it was not representative, didn't have it now (but still the low fps issue):

![0 78 1 180000 Re-Test](https://user-images.githubusercontent.com/63668523/147862288-c6b74a2e-ba5a-467a-bc74-662c640451f9.png)

I decided to compare the tab "details" of the task manager. Dosbox.exe consumes:

- 0.76 32bit: 3-7%
- 0.78.1 32bit: 10-13%

As you can see, no other process is consuming high CPU power (screenshot from 0.78.1 running):

![0 78 1 180000 CPU Re-Test](https://user-images.githubusercontent.com/63668523/147862310-5ef20f64-f5da-48f9-8544-2ab805e0f7b5.png)

So, not enough CPU power shouldn't be the problem, no? Because there is more then enough left, but the game is still stuttering. Could it be an issue with the graphic card or some driver issues? As you wrote that you have a smooth gaming experience even with 0.78.1, it could be something with my system (configuration), which only became a problem since version 0.77.1.

**Comment by Grounded0** [comment]
As DOSBox Staging is mostly a single thread program it matters how much a single thread is being stressed as thats where most of the action happens.

**Comment by kcgen** [comment]
Is there a way to toggle the NVidia GPU into game-mode? (I kid 😄 )

> So, not enough CPU power shouldn't be the problem, no? 

Adding to @Grounded0 's, the bulk of the CPU time goes into the dynamic core (the x86-64 core, authored by jmarsh) is already very fast. Threading it might actually slow it down due to excessive lock contention and each chunk being different sizes. So yes.. unfortunately throwing more cores at it won't help.  DOSBox likes few-and-fast instead of many-and-slow.

> it could be something with my system (configuration), which only became a problem since version 0.77.1.

The fact my my older, slower, i5 is producing the same FPS versus 0.76 (side-by-side Quake time demo results are identical), does have me agreeing here too. Can you try these builds: https://github.com/dosbox-staging/dosbox-staging/actions/runs/1646310582 , I've rolled back some of the Windows-specific cache memory protection calls. It could be that these were exasperating some (security?) features.

**Comment by RaymanXY** [comment]
OK you got it, at least for the x64 build!

- x64: Very smooth with cycles 100%, like the old 0.76 32bit and definitly smoother than the tested 0.79 Alpha 64bit from https://github.com/dosbox-staging/dosbox-staging/issues/1482#issuecomment-1003316927. With cycles = 180000 it's playable, but not so smooth. But hey, that solves the problem! I've tested Manic Karts as well and this is running also very good.

- x86: Crashes as soon as you go to the runway 😄 Sometimes it crashes before (the window is closing without error message). It doesn't matter if fullscreen or not. So I can't compare the 32bit version with each other. Note that the 3000 cycles in the title of the window is not representing the cycles which I had configured (always set to auto).
![x86 crash](https://user-images.githubusercontent.com/63668523/147961690-f037a3de-63c0-417c-a172-7acd51c43272.png)
